File: Python/getSkyline.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: getSkyline.py
@Function: 天际线问题
@Link: https://leetcode-cn.com/problems/the-skyline-problem/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-26
"""

from typing import List
import heapq


class Solution:
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        """顺序扫描优化"""
        s = []
        ans = []
        cur = 0
        for left, right, height in buildings:
            while s and s[0][1] < left:
                rh, r = heapq.heappop(s)
                if r >= cur:
                    if not s or rh != s[0][0]:
                        while s and r > s[0][1]:
                            heapq.heappop(s)
                        rh = -s[0][0] if s else 0
                        # 避免右端点重复
                        if r == ans[-1][0]:
                            ans[-1][1] = min(ans[-1][1], rh)
                        else:
                            ans.append([r, rh])
                    cur = r
            if not s or height > -s[0][0]:
                # 避免左端点重复的问题
                if ans and left == ans[-1][0]:
                    ans[-1][1] = height
                else:
                    ans.append([left, height])
            heapq.heappush(s, [-height, right])
        while s:
            rh, r = heapq.heappop(s)
            if r >= cur:
                if not s or rh != s[0][0]:
                    while s and r > s[0][1]:
                        heapq.heappop(s)
                    rh = -s[0][0] if s else 0
                    # 避免右端点重复
                    if r == ans[-1][0]:
                        ans[-1][1] = min(ans[-1][1], rh)
                    else:
                        ans.append([r, rh])
                cur = r
        return ans


# -------------------------
if __name__ == "__main__":
    buildings = [[2, 9, 10], [3, 7, 15], [
        5, 12, 12], [15, 20, 10], [19, 24, 8]]

    solution = Solution()
    get_sky_line = solution.getSkyline(buildings)
    print(f"The solution of this problem is {get_sky_line}")

File: C++/getSkyline.cpp
/**
 * @File    : getSkyline.cpp
 * @Brief   : 天际线问题
 * @Link    : https://leetcode-cn.com/problems/the-skyline-problem/
 * @Author  : Wei Li
 * @Date    : 2021-07-26
*/

#include <iostream>
#include <vector>
#include <utility>
#include <queue>

class Solution
{
public:
    std::vector<std::vector<int>> getSkyline(std::vector<std::vector<int>> &buildings)
    {
        std::vector<std::vector<int>> ans;
        std::priority_queue<std::pair<int, int>> max_heap; // <高度, 右端>
        int i = 0, len = buildings.size();
        int cur_x, cur_h;
        while (i < len || !max_heap.empty())
        {
            if (max_heap.empty() || i < len && buildings[i][0] <= max_heap.top().second)
            {
                cur_x = buildings[i][0];
                while (i < len && cur_x == buildings[i][0])
                {
                    max_heap.emplace(buildings[i][2], buildings[i][1]);
                    ++i;
                }
            }
            else
            {
                cur_x = max_heap.top().second;
                while (!max_heap.empty() && cur_x >= max_heap.top().second)
                {
                    max_heap.pop();
                }
            }
            cur_h = (max_heap.empty()) ? 0 : max_heap.top().first;
            if (ans.empty() || cur_h != ans.back()[1])
            {
                ans.push_back({cur_x, cur_h});
            }
        }
        return ans;
    }
};

template <typename T>
void DisplayElement(const T &container)
{
    for (auto element : container)
    {
        for (auto value : element)
        {
            std::cout << value << " ";
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

// ------------------------------
int main(int argc, char **argv)
{
    std::vector<std::vector<int>> buildings = {{2, 9, 10}, {3, 7, 15}, {5, 12, 12}, {15, 20, 10}, {19, 24, 8}};

    Solution solution;
    auto get_sky_line = solution.getSkyline(buildings);
    DisplayElement(get_sky_line);

    return 0;
}
